runtime/src/deterministic.rs

```diff
@@ -450,17 +495,48 @@ impl crate::Runner for Runner {
                 panic!("runtime stalled");
             }
             iter += 1;
-        }
+        };
+
+        // Stop the executor
+        executor.stop();
+
+        // Assert the context doesn't escape the start() function (behavior
+        // is undefined in this case)
+        assert!(
+            Arc::weak_count(&executor) == 0,
+            "executor still has weak references"
+        );
+
+        // Extract the executor from the Arc
+        let executor = Arc::into_inner(executor).expect("executor still has strong references");
+
+        (output, executor.checkpoint(storage))
```

Comment:
While I don't "love" that we have to provide storage as an argument here, I think it is cleaner than storing it on `Executor` (given it has nothing to do with task execution).